dynawo/sources/Models/Modelica/Dynawo/Electrical/Controls/Voltage/TCL.mo

```diff
@@ -23,62 +23,74 @@ model TCL "Tap Changer Lock (TCL)"
-    Boolean locked (start = locked0) "TCL locked ?";
+    State state(start = state0) "State of the automata throughout the simulation";
 
   equation
 
     // Check when the monitored voltage goes below UMin
-    // If the TCL is manually locked we ignore the voltage related events. Might be reviewed later.
-    when UMonitored.value < UMin and UMonitored.value > 0  and not (lockOrder.value) then
+    when UMonitored.value < UMin then
       UUnderMin = true;
       tUnderUmin = time;
-      Timeline.logEvent1(TimelineKeys.TapChangersArming);
-    elsewhen UMonitored.value > UMin and pre(UUnderMin) and not pre(lockOrder.value) then
+    elsewhen UMonitored.value >= UMin and pre(UUnderMin) then
       UUnderMin = false;
       tUnderUmin = Constants.inf;
-      Timeline.logEvent1(TimelineKeys.TapChangersUnarming);
     end when;
 
-    // Lock order activation
-    when ( (UUnderMin and time - tUnderUmin >= tLagBeforeLocked) or lockOrder.value ) and (not pre(locked)) then
-      locked = true;
+    // Automata state transitions
+    // Note that the deactivation from the Locked state is only manual
+    when UUnderMin and pre(state) == State. Standard then
+      state = State.Armed;
+      tLocked = Constants.inf;
+      Timeline.logEvent1(TimelineKeys.TapChangersArming);
+    elsewhen not(UUnderMin) and pre(state) == State.Armed then
+      state = State.Standard;
+      tLocked = Constants.inf;
+      Timeline.logEvent1(TimelineKeys.TapChangersUnarming);
+    elsewhen (((time - tUnderUmin >= tLagBeforeLocked) or (lockOrder.value and not(pre(lockOrder.value)))) and pre(state) == State.Armed) or (lockOrder.value and not(pre(lockOrder.value)) and pre(state) == State.Standard) then
+      state = State.Locked;
       tLocked = time;
       Timeline.logEvent1(TimelineKeys.TapChangersLocked);
-    elsewhen not(UUnderMin) and not(lockOrder.value) and (pre(locked)) then
-      locked = false;
+    elsewhen unlockOrder.value and not(pre(unlockOrder.value)) and pre(state) == State.Locked then
+      state = State.Standard;
       tLocked = Constants.inf;
       Timeline.logEvent1(TimelineKeys.TapChangersUnlocked);
     end when;
 
     // Lock order transmission to low voltage and high voltage tap-changers
-    when locked and time - tLocked >= tLagTransLockedT then
+    when time - tLocked >= tLagTransLockedT and pre(state) == State.Locked then
       lockedT = true;
       Timeline.logEvent1(TimelineKeys.TapChangersLockedT);
-    elsewhen not(locked) then
+    elsewhen unlockOrder.value and not(pre(unlockOrder.value)) then
       lockedT = false;
     end when;
 
-    when locked and time - tLocked >= tLagTransLockedD then
+    when time - tLocked >= tLagTransLockedD and pre(state) == State.Locked then
       lockedD = true;
       Timeline.logEvent1(TimelineKeys.TapChangersLockedD);
-    elsewhen not(locked) then
+    elsewhen unlockOrder.value and not(pre(unlockOrder.value)) then
       lockedD = false;
     end when;
```

Comment:
In the unlock parts, I would have expected similar conditions than the transition from State.Locked to State.Standard because it is the same mechanism.

Considering my previous remark, I would suggest having:
elsewhen unlockOrder.value and pre(state) = State.Locked then
   lockedD = false;
end when;

Reply:
In fact the lockedT and lockedB boolean should be reset when the transition from Locked to Standard appears. I updated the condition from:
`unlockOrder.value and not(pre(unlockOrder.value))`
to
`state == State.Standard and pre(state) == State.Locked`